pipelines/matrix/src/matrix/pipelines/integration/filters.py

```diff
@@ -63,7 +114,34 @@ def determine_most_specific_category(nodes: ps.DataFrame) -> ps.DataFrame:
         .withColumn("row_num", F.row_number().over(ps.Window.partitionBy("id").orderBy(F.col("depth").desc())))
         .filter(F.col("row_num") == 1)
         .drop("row_num")
-        .select("id", "category")
+        .select("id", F.col("candidate_category").alias("most_specific_from_all_categories"))
     )
 
-    return nodes.drop("category").join(mapping_table, on="id", how="left")
+    # Join with most specific categories
+    nodes_with_all_info = nodes_with_updated_categories.join(most_specific_mapping, on="id", how="left")
+
+    # Apply rules logic to determine final category
+    final_nodes = nodes_with_all_info.withColumn(
+        "final_category",
+        F.when(
+            # Rule 1: Core entities (have core_id) keep their existing category
+            F.col("core_id").isNotNull(),
+            F.col("category"),
+        )
+        .when(
+            # Rule 2: If all_categories only contains "biolink:NamedThing", keep existing category if more specific
+            (F.array_size("all_categories") == 1)
+            & (F.array_contains("all_categories", "biolink:NamedThing"))
+            & (F.col("category") != "biolink:NamedThing")
+            & (F.col("category").isNotNull()),
+            F.col("category"),
+        )
+        .otherwise(
+            # Rule 3: Use most specific from all_categories
+            F.coalesce("most_specific_from_all_categories", "category")
+        ),
+    )
+
+    return final_nodes.withColumn("category", F.col("final_category")).drop(
+        "most_specific_from_all_categories", "core_type", "final_category"
```

Comment:
Do we still need to drop `core_type`?
```suggestion
        "most_specific_from_all_categories"
```